src/Documentation/Frequently-Asked-Questions.md

```diff
@@ -21,64 +22,88 @@ Orleans has been production ready and used in production for several years.
+In case of a silo failure in the middle of a grain call you'll receive an exception that you can catch in your code and retry or do something else to handle the error according to your application logic.
+The grain that failed with the silo will get automatically re-instantiated upon a next call to it.
+The Orleans runtime does not eagerly recreate grains from a failed silo because many of them may not be needed immediately or at all.
+Instead, the runtime recreates such grains individually and only when a new request arrives for a particular grain.
+For each grain it picks one of the available silos as a new host.
+The benefit of this approach is that the recovery process is performed only for grains that are actually being used and it is spread in time and across all available silos, which improves the responsiveness of the system and the speed of recovery.
+Note also that there is a delay between the time when a silo fails and when the Orleans cluster detects the failure.
+The delay is a configurable tradeoff between the speed of detection and the probability of false positives.
+During this transition period all calls to the grain will fail, but after the detection of the failure the grain will be created, upon a new call to it, on another silo, so it will be eventually available.
+More information can be found [here](Runtime-Implementation-Details/Cluster-Management.md).
+
+## What happens if a grain call takes too long to execute?
+
+Since Orleans uses a cooperative multi-tasking model, it will not preempt the execution of a grain automatically but Orleans generates warnings for long executing grain calls so you can detect them.
+Cooperative multi-tasking has a much better throughput compared to preemptive multi-tasking.
+Keep in mind that grain calls should not execute any long running tasks like IO operations synchronously and should not block on other tasks to complete.
+All waiting should be done asynchronously using the `await` keyword or other asynchronous waiting mechanisms.
+Grains should return as soon as possible to let other grains execute for maximum throughput.
 
 ## In what cases can a split brain (same grain activated in multiple silos at the same time) happen?
 
-This can never happen during normal operations and each grain will have one and only one instance per ID.
-The only time this can occur is when a silo crashes or if it's killed without being allowed to properly shutdown.
-In that case there is a 30-60 seconds window (based on configuration) where a grain can exist in multiple silos before one is removed from the system.
+During normal operations the Orleans runtime guarantees that each grain will have at most one instance in the cluster.
+The only time this guarantee can be violated is when a silo crashes or gets killed without a proper shutdown.
+In that case, there is a ~30 second (based on configuration) window where a grain can potentially get temporarily instantiated in more than one silo.
 
```

Comment:
I don't think this fits the FAQ (at least in its current form). Plus, it won't necessarily get immediately deactivated because grain code may catch the exception.

In general, I think we need to rewrite the FAQ altogether. But that's not what I tried to do here. I tried only to update it and fix inaccuracies.